GeoTIFF output

tjlang opened the pull request:
Added GeoTIFF output capabilities to `pyart.io` module. New function is `pyart.io.write_grid_geotiff`, and it accepts Py-ART `Grid` objects. New functionality adds `gdal` as an optional dependency. If not installed, the function will complain and not work.

Example usage:
```
pyart.io.write_grid_geotiff(
    'test_radar_fp.tiff', grid, field='ZC', cmap='pyart_NWSRef',
    rgb=True, vmin=0, vmax=75, warp=True, sld=False)
```

This creates a simple 3-channel RGB GeoTIFF of composite reflectivity (ZC field), warped to a WGS84 lat/lon grid, using the `pyart_NWSRef` color table with min/max set at (0, 75). Here is what the output looks like in GIS software:
<img width="1042" alt="screen shot 2016-03-23 at 4 04 35 pm" src="https://cloud.githubusercontent.com/assets/3036387/14001550/78ca2d5c-f114-11e5-89ba-f486e6164f3a.png">

One can also output single-channel floating point data to the GeoTIFF, and create an ancillary Style Layer Descriptor (SLD) file containing color table info.

Example usage:
```
import dualpol
hidcolor = dualpol.HidColors()
pyart.io.write_grid_geotiff(
    'test_radar.tiff', grid, field='FH', level=4, cmap=hidcolor.cmaphid,
    rgb=False, vmin=0, vmax=10, color_levels=10, warp=True, sld=True)
```

This creates a GeoTIFF with HID index values following the `dualpol` paradigm (https://github.com/nasa/DualPol). The SLD file that is created contains the default `dualpol` HID colormap, which is categorical and has 10 entries. The output is valid for vertical level 4 in the Py-ART `Grid` object (2.5 km MSL in this case).

What it looks like in GIS software:
<img width="1042" alt="screen shot 2016-03-23 at 4 36 50 pm" src="https://cloud.githubusercontent.com/assets/3036387/14001740/9f71b10e-f115-11e5-917b-ceb8f6f50df8.png">

jjhelmus commented:
Very cool.  I'm about to leave for the day but will look at this in detail tomorrow.

jjhelmus commented on pyart/io/output_to_geotiff.py:
```diff
@@ -0,0 +1,246 @@
+        gridvol = filename
+    else:
+        fname = filename
+        gridvol = grid
```
I understand the idea behind this but I do not like it.  Either a user gives the parameter in the correct order or the function should fail.

tjlang replied:
Function now will fail if arguments are incorrect. The order of required arguments now is more logical, first `grid`, then `filename`, then `field`.

jjhelmus commented on pyart/io/output_to_geotiff.py:
```diff
@@ -0,0 +1,246 @@
+    Parameters
+    ----------
+    filename : str
+        Filename to save grid to
```
This is not really the filename written to since the field name, etc is added to the name.  Perhaps specifying a template would be better.

tjlang replied:
Now refer to `filename` argument as a filename template. The function also can now handle it if the user did not add a suffix.

jjhelmus commented on pyart/io/output_to_geotiff.py:
```diff
@@ -0,0 +1,246 @@
+                os.system('gdalwarp -t_srs \'+proj=longlat +ellps=WGS84 ' +
+                          '+datum=WGS84 +no_defs\' ' + ofile + ' ' +
+                          ofile + '_tmp.tif')
+                os.system('mv ' + ofile + '_tmp.tif' + ' ' + ofile)
```
This is not cross platform, `shutil.move` should work on more (all?) platforms.

tjlang replied:
Now use `shutil.move`.

jjhelmus commented on pyart/io/output_to_geotiff.py:
```diff
@@ -0,0 +1,246 @@
+    name, end = fname.split('.')
+
+    f_flag = False
+    for f, field_dic in gridvol.fields.iteritems():
```
iteritems does not work in Python 3, use items.

tjlang replied:
Removed the loop and now simply examine whether `field` appears in the dictionary's keys.

jjhelmus commented on pyart/io/output_to_geotiff.py:
```diff
@@ -0,0 +1,246 @@
+    IMPORT_FLAG = False
+
+
+def write_grid_geotiff(filename, grid, field=None, rgb=False, level=None,
```
There is no documentation of the field parameter in the docstring.

tjlang replied:
Made `field` a required argument. Also reordered the requirement arguments to make them more logical in order.

jjhelmus commented on pyart/io/output_to_geotiff.py:
```diff
@@ -0,0 +1,246 @@
+            # print("rangestep :" + str(rangestep))
+            lat = gridvol.axes['lat']['data'][0]
+            lon = gridvol.axes['lon']['data'][0]
+            filled = gridvol.fields[field]['data'].filled(fill_value=-32768)
```
This does not work if the field data is not a masked array.

tjlang replied:
Added logic that handles the case of a standard unmasked `ndarray`.

jjhelmus commented on pyart/io/output_to_geotiff.py:
```diff
@@ -0,0 +1,246 @@
+            ofile = name + "_" + f + "." + end
+            nz, ny, nx = gridvol.fields[field]['data'].shape
+            rcells = ny * 0.5
+            dist = max(gridvol.axes['x_disp']['data'])
```
axes is deprecated, use grid.x['data']

tjlang replied:
Fixed.

jjhelmus commented on pyart/io/output_to_geotiff.py:
```diff
@@ -0,0 +1,246 @@
+        False - Don't do this
+    """
+    if not IMPORT_FLAG:
+        print('GDAL not detected, GeoTIFF output failure!')
```
This should raise a MissingOptionalDependency exception like other functions in Py-ART which depend on an optional dependency.

tjlang replied:
Added this capability.

jjhelmus commented on pyart/io/output_to_geotiff.py:
```diff
@@ -0,0 +1,246 @@
+                          ofile + '_tmp.tif')
+                os.system('mv ' + ofile + '_tmp.tif' + ' ' + ofile)
+    if not f_flag:
+        print('Failed - field not found.')
```
This should raise an exception.

tjlang replied:
A `KeyError` exception is now raised if the `field` doesn't exist in the `Grid` object.

jjhelmus commented on pyart/io/output_to_geotiff.py:
```diff
@@ -0,0 +1,246 @@
+        color_levels = 255
+    name, end = filename.split('.')
+    ofile = name + '.sld'
+    fileobj = open(ofile, 'w')
```
This needs to be closed or the writing should be done in a context manager.

jjhelmus commented:
@tjlang This is a great start, and I'd like to get this included in Py-ART but it needs a bit or work.  I added a number of in-line comments on the code to begin with.  The docstrings will also need to be updated to match the [NumPy documentation standards](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt) and the functions will need to be added to the module level docstrings.

tjlang commented:
I addressed the requested revisions and also updated docstrings. Please let me know if you see anything else that needs to be addressed before merging.

jjhelmus commented on pyart/io/output_to_geotiff.py:
```diff
@@ -0,0 +1,318 @@
+"""
+pyart.io.write_grid_geotiff
+====================
```
This should extend to the end of the text above

tjlang replied:
Done

jjhelmus commented on pyart/io/output_to_geotiff.py:
```diff
@@ -0,0 +1,318 @@
+               extension.
+        False - Don't do this.
+
+    Returns
```
This section can be removed.  The convention is that the lack of a Return section indicates that the function does not return anything or more precisely returns None.

tjlang replied:
Removed

jjhelmus commented on pyart/io/output_to_geotiff.py:
```diff
@@ -0,0 +1,318 @@
+        raise MissingOptionalDependency(
+            'GDAL not detected, GeoTIFF output failure!')
+
+    if field in grid.fields.keys():
```
This would be more readable if the failure is placed first.  Also less indentation is needed.

``` Python
    if field not in grid.fields.keys():
        raise KeyError('Failed -', field, 'field not found in Grid object.')

    # Determine whether filename template already contains a suffix
    ...
```

tjlang replied:
Done

jjhelmus commented on pyart/io/output_to_geotiff.py:
```diff
@@ -0,0 +1,318 @@
+        else:
+            name = filename
+            end = 'tif'
+        ofile = name + "_" + field + "." + end
```
What is the reasoning behind building the filename from the user provided string and the field name?  Could we just use the filename that the user provides to the function instead of building it from a template?  
I'm not against using a template but none of the other functions that write files in Py-ART use this setup.

tjlang replied:
Changed to not use the field in the filename. If there is not a suffix, however, '.tif' is appended so that `_create_sld` gets a suffix it can change to '.sld'

jjhelmus commented on pyart/io/output_to_geotiff.py:
```diff
@@ -0,0 +1,318 @@
+        rangestep = grid.x['data'][1] - grid.x['data'][2]
+        lat = grid.origin_latitude['data'][0]
+        lon = grid.origin_longitude['data'][0]
+        # Check if masked array; if so, fill missing data
```
I think `np.ma.filled(grid.fields[field]['data'], fill_value=-32768)` will accomplish the same thing.

tjlang replied:
OK, trying that instead of the if block

jjhelmus commented on pyart/io/output_to_geotiff.py:
```diff
@@ -0,0 +1,318 @@
+        data = data.astype(float)
+        try:
+            data[data == -32768] = np.nan
+        except:
```
What exception is being caught here?  Is it safe to ignore it?

tjlang replied:
I seem to recall some weird error during development, but I can't recall what it was. I took this exception handling block out and it seems to be working OK. If we start getting errors in the future we can add this block back in but be more specific about the exception being raised.

jjhelmus commented on pyart/io/__init__.py:
```diff
@@ -78,6 +78,7 @@
 from .uf import read_uf
 from .uf_write import write_uf
 from .grid_io import read_grid, write_grid, read_legacy_grid
+from .output_to_geotiff import write_grid_geotiff
```
Add the `write_grid_geotiff` function in one of the documentation sections at the top of this file.  That way the function is included in the User Guide

tjlang replied:
Done

jjhelmus commented:
Thanks for the updates, looking really good.  I made a few  few minor in code comments.  

Do you have any tests for this new functions that could be included?  The following will at least test that the function outputs to a file that contains something.  Let me know if you want to address this, I'm happy to write a few tests for the module if you would prefer.

``` Python
""" Unit Tests for Py-ART's output_to_geotiff.py module. """

import pyart


def test_write_grid_geotiff():
    grid = pyart.testing.make_target_grid()
    with pyart.testing.InTemporaryDirectory():
        pyart.io.write_grid_geotiff(grid, 'test.foo', 'reflectivity')
        # check that something was written to the file
        with open('test_reflectivity.foo', 'rb') as f:
            assert len(f.read()) > 0
```

jjhelmus commented on pyart/io/output_to_geotiff.py:
```diff
@@ -0,0 +1,318 @@
+    Other Parameters
+    ----------------
+    rbg : bool, optional
+        True - Output 3-band RGB GeoTIFF
```
Sphinx doesn't like something about these True/False blocks and is giving a warning.  I think a blank line is needed after each option, one after the True block, one after the False, but I'm not sure.  Easiest method I know to check if these work is to try building the docs.  I can do this before merging if you would prefer.

tjlang replied:
Added blank spaces below each True and False text block

tjlang commented:
OK, further edits have been made. I did delete the appending of the field to the filename, so you'll need to fix your test function to account for that. I'll try to think about additional test functions, but if you have some ideas feel free to put them wherever they are supposed to go. Thanks!

Everything seems to be working. Here's a local multi-radar composite reflectivity mosaic from tonight, with all hell breaking loose around HSV.

![Uploading Screen Shot 2016-03-31 at 7.37.36 PM.png…]()

jjhelmus commented:
Great.  Everything seems in order.  I'll create another PR adding some basic tests for the new functionality.  Thanks for seeing this through @tjlang.
